### PR TITLE
Add kpro_connection:sasl_reauthenticate_after/2

### PR DESCRIPTION
**Pull request by urmastalimaa**
Allows authentication callback adapters to re-authenticate the
connection before session lifetime expires. An example use case is given
under function documentation.

This change requires adding the connection config, which contains SASL
options to the state record.

In theory, the connection process could read `session_lifetime_ms` out
of the SASL authentication response and set a timer itself, but this
would require quite a few changes to existing interfaces as the sasl
authentication functions return just ok and such an addition would
probably need to be behind a feature flag.

### CODE REVIEW

File: src/kpro_connection.erl
Comment (oskarkook) [inline src/kpro_connection.erl:189]: It will depend on the taste of the maintainers of `kafka_protocol` ultimately, but this could also just be `sasl_reauthenticate` and leave the timer-ing to the caller.

**Review by oskarkook — APPROVED** [review]
Legendary. I will be telling tales how Urmas fixed the Kafka client.

File: src/kpro_connection.erl
Comment (oskarkook) [reply src/kpro_connection.erl:189]: Timering in the caller does seem very inconvenient, so disregard the comment 😄

**Review by indrekj — APPROVED** [review]
you'll probably get a better review from the author of the library

File: src/kpro_connection.erl
Comment (urmastalimaa) [reply src/kpro_connection.erl:189]: Yes, precisely because it needs to send a precise message, as the message atom is part of the private API.
So it would need to be `erlang:send_after(Time, Connection, kpro_connection:make_sasl_reauthenticate_msg())`

**Comment by urmastalimaa** [comment]
I will make a PR to the upstream when we have moved some of our services over and validated this

**Pull request merged**
